// @vue/language-plugin-pug@1.7.0/out/index.js
"use strict";
const pug = require("volar-service-pug/out/languageService");
const source_map_1 = require("@volar/source-map");
const plugin = ({ modules }) => {
    return {
        name: require('../package.json').name,
        version: 1,
        compileSFCTemplate(lang, template, options) {
            if (lang === 'pug') {
                const pugFile = pug?.baseParse(template);
                const map = new source_map_1.SourceMap(pugFile.mappings);
                if (pugFile) {
                    const compiler = modules['@vue/compiler-dom'];
                    const completed = compiler.compile(pugFile.htmlCode, {
                        ...options,
                        onWarn(warning) {
                            options?.onWarn?.(createProxyObject(warning));
                        },
                        onError(error) {
                            options?.onError?.(createProxyObject(error));
                        },
                    });
                    return createProxyObject(completed);
                    function createProxyObject(target) {
                        return new Proxy(target, {
                            get(target, prop) {
                                if (prop === 'offset') {
                                    const htmlOffset = target.offset;
                                    for (const mapped of map.toSourceOffsets(htmlOffset)) {
                                        return mapped[0];
                                    }
                                    return -1;
                                }
                                const value = target[prop];
                                if (typeof value === 'object') {
                                    return createProxyObject(target[prop]);
                                }
                                return value;
                            }
                        });
                    }
                }
            }
        },
    };
};
module.exports = plugin;
