// @vite-pwa/vitepress@0.2.3/dist/index.cjs
'use strict';

const vitePluginPwa = require('vite-plugin-pwa');

function configurePWAOptions(options) {
  if (!options.outDir)
    options.outDir = ".vitepress/dist";
  if (options.strategies === "injectManifest") {
    options.injectManifest = options.injectManifest ?? {};
  } else {
    options.workbox = options.workbox ?? {};
    if (options.registerType === "autoUpdate" && (options.injectRegister === "script" || options.injectRegister === "inline")) {
      options.workbox.clientsClaim = true;
      options.workbox.skipWaiting = true;
    }
  }
}

function withUserConfig(config) {
  let viteConf = config.vite;
  if (!viteConf) {
    viteConf = {};
    config.vite = viteConf;
  }
  let vitePlugins = viteConf.plugins;
  if (typeof vitePlugins === "undefined") {
    vitePlugins = [];
    viteConf.plugins = vitePlugins;
  }
  if (vitePlugins && vitePlugins.length > 0) {
    const pwaPlugin = vitePlugins.find((i) => i && typeof i === "object" && "name" in i && i.name === "vite-plugin-pwa");
    if (pwaPlugin)
      throw new Error("Remove vite-plugin-pwa plugin from Vite Plugins entry in VitePress config file");
  }
  const { pwa = {} } = config;
  configurePWAOptions(pwa);
  let api;
  vitePlugins.push(
    vitePluginPwa.VitePWA({ ...pwa }),
    {
      name: "vite-plugin-pwa:vitepress",
      apply: "build",
      enforce: "post",
      configResolved(resolvedViteConfig) {
        if (!resolvedViteConfig.build.ssr)
          api = resolvedViteConfig.plugins.find((p) => p.name === "vite-plugin-pwa")?.api;
      }
    }
  );
  const vitePressConfig = config;
  const userTransformHead = vitePressConfig.transformHead;
  const userBuildEnd = vitePressConfig.buildEnd;
  vitePressConfig.transformHead = async (ctx) => {
    const head = await userTransformHead?.(ctx) ?? [];
    const webManifestData = api?.webManifestData();
    if (webManifestData) {
      const href = webManifestData.href;
      if (webManifestData.useCredentials)
        head.push(["link", { rel: "manifest", href, crossorigin: "use-credentials" }]);
      else
        head.push(["link", { rel: "manifest", href }]);
    }
    const registerSWData = api?.registerSWData();
    if (registerSWData && registerSWData.shouldRegisterSW) {
      if (registerSWData.inline) {
        head.push([
          "script",
          { id: "vite-plugin-pwa:inline-sw" },
          `if('serviceWorker' in navigator) {window.addEventListener('load', () => {navigator.serviceWorker.register('${registerSWData.inlinePath}', { scope: '${registerSWData.scope}' })})}`
        ]);
      } else {
        head.push([
          "script",
          {
            id: "vite-plugin-pwa:register-sw",
            src: registerSWData.registerPath
          }
        ]);
      }
    }
    return head;
  };
  let allowlist;
  if (pwa.strategies !== "injectManifest" && pwa.experimental?.includeAllowlist === true) {
    pwa.workbox = pwa.workbox ?? {};
    allowlist = pwa.workbox.navigateFallbackAllowlist = pwa.workbox.navigateFallbackAllowlist ?? [];
    pwa.workbox.runtimeCaching = pwa.workbox.runtimeCaching ?? [];
    pwa.workbox.runtimeCaching.push({
      urlPattern: ({ request, sameOrigin }) => {
        return sameOrigin && request.mode === "navigate";
      },
      handler: "NetworkOnly",
      options: {
        plugins: [{
          /* this callback will be called when the fetch call fails */
          handlerDidError: async () => Response.redirect("404", 302),
          /* this callback will prevent caching the response */
          cacheWillUpdate: async () => null
        }]
      }
    });
  }
  vitePressConfig.buildEnd = async (siteConfig) => {
    await userBuildEnd?.(siteConfig);
    if (api && !api.disabled) {
      if (typeof allowlist !== "undefined") {
        const base = siteConfig.site.base ?? "/";
        for (const page of siteConfig.pages) {
          const regex = page === "index.md" ? escapeStringRegexp(base) : escapeStringRegexp(`${base}${page.replace(/\.md$/, "")}`);
          allowlist.push(new RegExp(`^${regex}(\\.html)?$`));
        }
      }
      await api.generateSW();
    }
  };
  return vitePressConfig;
}
function escapeStringRegexp(value) {
  return value.replace(/[|\\{}()[\]^$+*?.]/g, "\\$&").replace(/-/g, "\\x2d");
}

async function withPwa(config) {
  if (typeof config === "function") {
    return async (ctx) => {
      const userConfig = await config(ctx);
      return withUserConfig(userConfig);
    };
  }
  return withUserConfig(await config);
}

exports.withPwa = withPwa;
